### DISCUSSION TITLE
apply initial stress fields on a 2D mesh boundary in form of expressions

### DISCUSSION THREAD

**Discussion by Aayush-T** [idaholab/moose#28812] [Q&A Modules: Solid mechanics]
Hi Everyone,
So, I am trying to apply an initial force condition to one of the surfaces of mesh in my simulations. This example that I made for now is for a 2D case; I will modify it for 3D.

I am working with a 2D mesh and I need to apply initial stress fields (σ_xx, σ_yy, σ_xy) on the right boundary. These stress fields are expressed as functions of the length of the mesh (i.e., vary along the boundary).

I've been unable to find an example that demonstrates how to specify these stress fields correctly. The closest example I found was in issue https://github.com/idaholab/moose/discussions/26104 , but I still couldn't figure out how to properly define and apply the stress expressions on the boundary.

To give an idea, this is my mesh.
![Screenshot from 2024-10-09 17-24-04](https://github.com/user-attachments/assets/b29dbe49-7a51-4c11-aced-f11119267fde)

One method I tried to think was to get reaction forces from these stresses and then apply these forces as the function of length over mesh in the boundary condition. Is this correct?

and this is my code:
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  elem_type = QUAD4
  displacements = 'disp_x disp_y'
  nx = 10
  ny = 10
[]

[Variables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
[]

[GlobalParams]
  volumetric_locking_correction = true
[]

[AuxVariables]
[./tdisp]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./stress_yy]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./e_yy]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./fp_yy]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./gss]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./euler1]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./euler2]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./euler3]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Functions]
  [./tdisp]
    type = ParsedFunction
    expression = 0.01*t
  [../]
[]

[UserObjects]
  [./prop_read]
    type = PropertyReadFile
    prop_file_name = 'euler_ang_file_2.txt'
    # Enter file data as prop#1, prop#2, .., prop#nprop
    nprop = 3
    read_type = element
  [../]
[]

[AuxKernels]
  [./stress_yy]
    type = RankTwoAux
    variable = stress_yy
    rank_two_tensor = stress
    index_j = 1
    index_i = 1
    execute_on = timestep_end
  [../]
  [./e_yy]
    type = RankTwoAux
    variable = e_yy
    rank_two_tensor = lage
    index_j = 1
    index_i = 1
    execute_on = timestep_end
  [../]
  [./fp_yy]
    type = RankTwoAux
    variable = fp_yy
    rank_two_tensor = fp
    index_j = 1
    index_i = 1
    execute_on = timestep_end
  [../]
  [./gss]
    type = MaterialStdVectorAux
    variable = gss
    property = state_var_gss
    index = 0
    execute_on = timestep_end
  [../]
  [./euler1]
    type = MaterialRealVectorValueAux
    variable = euler1
    property = Euler_angles
    component = 0
    execute_on = timestep_end
  [../]
  [./euler2]
    type = MaterialRealVectorValueAux
    variable = euler2
    property = Euler_angles
    component = 1
    execute_on = timestep_end
  [../]
  [./euler3]
    type = MaterialRealVectorValueAux
    variable = euler3
    property = Euler_angles
    component = 2
    execute_on = timestep_end
  [../]
[]

[BCs]
  [./fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]
  [./fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]

  [./tdisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = tdisp
  [../]
  [./pre_load]
    type = FunctionNeumannBC
    variable = disp_x
    boundary = right
    function = -50*y #This is for F_x. Similarly, we can do variable =  disp_y for F_y
  [../]
[]

[UserObjects]
  [./slip_rate_gss]
    type = CrystalPlasticitySlipRateGSS
    variable_size = 12
    slip_sys_file_name = input_slip_sys.txt
    num_slip_sys_flowrate_props = 2
    flowprops = '1 4 0.001 0.1 5 8 0.001 0.1 9 12 0.001 0.1'
    uo_state_var_name = state_var_gss
  [../]
  [./slip_resistance_gss]
    type = CrystalPlasticitySlipResistanceGSS
    variable_size = 12
    uo_state_var_name = state_var_gss
  [../]
  [./state_var_gss]
    type = CrystalPlasticityStateVariable
    variable_size = 12
    groups = '0 4 8 12'
    group_values = '60.8 60.8 60.8'
    uo_state_var_evol_rate_comp_name = state_var_evol_rate_comp_gss
    scale_factor = 1.0
  [../]
  [./state_var_evol_rate_comp_gss]
    type = CrystalPlasticityStateVarRateComponentGSS
    variable_size = 12
    hprops = '1.0 541.5 109.8 2.5'
    uo_slip_rate_name = slip_rate_gss
    uo_state_var_name = state_var_gss
  [../]
[]

[Materials]
  [./crysp]
    type = FiniteStrainUObasedCP
    stol = 1e-2
    tan_mod_type = exact
    uo_slip_rates = 'slip_rate_gss'
    uo_slip_resistances = 'slip_resistance_gss'
    uo_state_vars = 'state_var_gss'
    uo_state_var_evol_rate_comps = 'state_var_evol_rate_comp_gss'
  [../]
  [./strain]
    type = ComputeFiniteStrain
    displacements = 'disp_x disp_y'
  [../]
  [./elasticity_tensor]
    type = ComputeElasticityTensorCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
    read_prop_user_object = prop_read
  [../]
[]

[Postprocessors]
[./tdisp]
    type = ElementAverageValue
    variable = tdisp
  [../]
  [./stress_yy]
    type = ElementAverageValue
    variable = stress_yy
  [../]
  [./e_yy]
    type = ElementAverageValue
    variable = e_yy
  [../]
  [./fp_yy]
    type = ElementAverageValue
    variable = fp_yy
  [../]
  [./gss]
    type = ElementAverageValue
    variable = gss
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  dt = 0.01
  solve_type = 'PJFNK'

  petsc_options_iname = -pc_hypre_type
  petsc_options_value = boomerang
  nl_abs_tol = 1e-10
  nl_rel_step_tol = 1e-10
  dtmax = 10.0
  nl_rel_tol = 1e-10
  end_time = 1
  dtmin = 0.01
  num_steps = 100
  nl_abs_step_tol = 1e-10
[]

[Outputs]
  exodus = true
  csv = true
[]

[Kernels]
  [./TensorMechanics]
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
  [../]
[]
```

**Comment by GiudGiud** [comment]
Hello

This can be done by using Functions to create the boundary stress values, since they are only functions of space and are pre-determined at the beginning of the simulation.
When solving the solid mechanics equations, these stresses can be applied using the Pressure action for boundary conditions
https://mooseframework.inl.gov/syntax/BCs/Pressure/index.html
with the function parameter

There is also more documentation here
https://mooseframework.inl.gov/source/bcs/Pressure.html

I am actually not sure how to apply sigma_xy. This should work for xx and yy.
@gambka do you know?

**Reply by Aayush-T** [reply]
Hi @GiudGiud 
Thanks for the response.

Till now, I have been unsuccesful in getting it done for stress_xy and stress_yz.
I thought of this approach. I would like to know your thoughts on this:
Let us say that I somehow obtain the force distribution F_x, F_y, F_z from these stresses and apply them like this:

[./pre_load_x]
    type = FunctionNeumannBC
    variable = disp_x
    boundary = right
    function = -50*y + 40*z #This is for F_x. Similarly, we can do variable =  disp_y for F_y
  [../]

[./pre_load_y]
    type = FunctionNeumannBC
    variable = disp_y
    boundary = right
    function = -40*x +30*y
  [../]

[./pre_load_z]
    type = FunctionNeumannBC
    variable = disp_y
    boundary = right
    function = -50*z + 23*y 
  [../]
  What do you think of this implementation?

**Reply by GiudGiud** [reply]
I think this are still stress_xx/yy/zz terms, they just depends spatially on the two other dimensions
@recuero do you know how to apply off-diagonal stresses at the boundary?

**Reply by Aayush-T** [reply]
[./pre_load_z]
type = FunctionNeumannBC
variable = disp_y
boundary = right
function = -50z + 23y
[../]

Is it not the load (in a way Force?)

**Reply by maxnezdyur** [reply]
Are we sure that 

```
[pre_load_y]
type = FunctionNeumannBC
variable = disp_y
boundary = right
function = -40x +30y
[../]
```
is not applying the sigma_xy stress instead? I don't think you can apply sigma_yy on the right boundary when the normals are [1,0,0].

For 2d, if we want to apply sigma_xx, sigma_yy, sigma_xy on a surface of normal (nx,ny), you would do T = sigma dot n. Then apply T_x and T_y, when the normal is [1,0] then T_y = sigma_xy * 1 + simga_yy * 0.

**Reply by GiudGiud** [reply]
ah good point. I didnt think that the normal was not aligned

**Reply by Aayush-T** [reply]
So, I have to eventually apply it for 3D. So, technically it is based on surface roughness but, here, I can assume that the normals are aligned.
Based on what you have explained above, I came up with this:
@GiudGiud @maxnezdyur 
Right Face in 3D:
Normal vector will be: n = [1,0,0] (n_x = 1, n_y = 0, n_z = 0)
and thus traction components will be:
T_x = sigma_xx * 1 + sigma_xy * 0 + sigma_xz * 0
T_y = sigma_xy * 1 + sigma_yy * 0 + sigma_yz * 0
T_z = sigma_xz * 1 + sigma_yz * 0 + sigma_zz * 0

and thus the overall stress map will be:

[BCs]
  [./pre_load_x]
    type = FunctionNeumannBC
    variable = disp_x
    boundary = right
    function = T_x
  [../]
  [./pre_load_y]
    type = FunctionNeumannBC
    variable = disp_y
    boundary = right
    function = T_y
  [../]
  [./pre_load_z]
    type = FunctionNeumannBC
    variable = disp_z
    boundary = right
    function = T_z
  [../]
[]
